master/static/srv/ship_logs.py

```diff
@@ -377,6 +398,89 @@ def ship_special(self, msg: str, metadata: Dict[str, str], emit_stdout_logs: boo
         backoffs = [0, 1, 5, 10, 15]
         self.ship(data, backoffs)
 
+    def assert_master_is_reachable(self):
+        """
+        Before we start actually shipping logs, try to confirm that we can reach the master at all.
+
+        If we can't, log an overly-detailed explanation to help sysadmins debug their cluster.
+
+        Unfortunately, if we can't reach the master, they'll have to use the /ship_logs escape hatch
+        to see this message, or be on slurm where we pull slurms' stderr into task logs
+        automatically.
+        """
+        backoffs = [0, 1, 5, 10, 15]
+        for delay in backoffs:
+            time.sleep(delay)
+            if self.try_reaching_master():
+                return
+        raise RuntimeError("failed to ever reach master")
+
+    def try_reaching_master(self) -> bool:
+        endpoint = "/api/v1/me"
+        url = f"{self.base_url}{endpoint}"
+        req = urllib.request.Request(url, headers=self.headers)
+        try:
+            with urllib.request.urlopen(req, context=self.context) as resp:
+                _ = resp.read()
+            return True
+        except Exception as e:
+            # urllib stacktraces are awful, so in cases where we can tell what happened, avoid
+            # printing it at all.
+            exc_info = True
+            detail = (
+                "This may be due to an address resolution problem, a certificate problem, a "
+                "firewall problem, a proxy problem, or some other networking error."
+            )
+            if isinstance(e, urllib.error.HTTPError):
+                detail = f"GET {url} returned status code: {e.code} and reason: {e.reason}."
+                exc_info = False
+            elif isinstance(e, urllib.error.URLError):
+                if isinstance(e.reason, socket.gaierror):
+                    detail = f"Name lookup for {self.master_url} failed: {e.reason}"
+                    exc_info = False
+                elif isinstance(e.reason, ConnectionRefusedError):
+                    detail = f"The connection to {self.master_url} was refused."
+                    exc_info = False
+                elif isinstance(e.reason, ssl.SSLCertVerificationError):
+                    detail = f"There was a tls verification error: {e.reason.verify_message}."
+                    exc_info = False
+
+            noverify = self.cert_file.lower() == "noverify"
+            if not noverify:
+                cert_content = None
+            else:
+                with open(self.cert_file, "r") as f:
```

Comment:
I should have said something on the earlier review, but there should probably be some error checking before we get to here. Maybe `shipper` should take contents rather than a filepath.

Reply:
Ok, after thinking more about it, I think it would be pretty annoying for every caller of Shipper to have to read the file, when every caller of shipper expects a file input instead of content input.

But I would be happy to read the cert content inside `Shipper.__init__`.  That's more robust anyway, because I think the SSLContext might puke if you pass it a file that doesn't actually exist.